File: exercism/python/queen_attack.py
class Queen:
    def __init__(self, row, column):
        if row < 0:
            raise ValueError("row not positive")
        if row > 7:
            raise ValueError("row not on board")
        if column < 0:
            raise ValueError("column not positive")
        if column > 7:
            raise ValueError("column not on board")
        self.row = row
        self.column = column

    def can_attack(self, another_queen):
        r = self.row == another_queen.row
        c = self.column == another_queen.column
        if r and c:
            raise ValueError("Invalid queen position: both queens in the same square")
        return abs(self.column - another_queen.column) == abs(self.row - another_queen.row) or r or c

File: exercism/rust/queen-attack.rs
#[derive(Debug)]
pub struct ChessPosition {
    rank: i32,
    file: i32
}

#[derive(Debug)]
pub struct Queen(ChessPosition);

impl ChessPosition {
    pub fn new(rank: i32, file: i32) -> Option<Self> {
        if 0 > rank || rank > 7 || 0 > file || file > 7 {
            return None;
        }
        return Some(ChessPosition{rank, file});
    }
}

impl Queen {
    pub fn new(position: ChessPosition) -> Self {
        Self(position)
    }

    pub fn can_attack(&self, other: &Queen) -> bool {
        self.0.rank == other.0.rank ||
        self.0.file == other.0.file ||
        (self.0.rank- other.0.file).abs() == (self.0.file - other.0.file).abs()
    }
}
